pennylane/beta/tapes/tape.py

```diff
@@ -0,0 +1,776 @@
+            qml.RY(0.543, wires=0)
+            qml.CNOT(wires=[0, 'a'])
+            qml.RX(0.133, wires='a')
+            expval(qml.PauliZ(wires=[0]))
+
+    Once constructed, information about the quantum circuit can be queried:
+
+    >>> tape.operations
+    [RX(0.432, wires=[0]), RY(0.543, wires=[0]), CNOT(wires=[0, 'a']), RX(0.133, wires=['a'])]
+    >>> tape.observables
+    [expval(PauliZ(wires=[0]))]
+    >>> tape.get_parameters()
+    [0.432, 0.543, 0.133]
+    >>> tape.wires
+    <Wires = [0, 'a']>
+    >>> tape.num_params
+    3
+
+    The :class:`~.beta.tapes.NewCircuitGraph` can also be accessed:
+
+    >>> tape.graph
+    <pennylane.beta.tapes.circuit_graph.NewCircuitGraph object at 0x7fcc0433a690>
+
+    Once constructed, the quantum tape can be executed directly on a supported
+    device:
+
+    >>> dev = qml.device("default.qubit", wires=[0, 'a'])
+
+    Execution can take place either using the in-place constructed parameters,
+    >>> tape.execute(dev)
+    [0.77750694]
+
+    or by providing parameters at run time:
+
+    >>> tape.execute(dev, params=[0.1, 0.1, 0.1])
+    [0.99003329]
+
+    The Jacobian can also be computed using finite difference:
+
+    >>> tape.jacobian(dev)
+    [[-0.35846484 -0.46923704  0.        ]]
+    >>> tape.jacobian(dev, params=[0.1, 0.1, 0.1])
+    [[-0.09933471 -0.09933471  0.        ]]
+
+    Finally, the trainable parameters can be explicitly set, and the values of
+    the parameters modified in-place:
+
+    >>> tape.trainable_params = {0} # set only the first parameter as free
+    >>> tape.set_parameters(0.56)
+    >>> tape.get_parameters()
+    [0.56]
+    >>> tape.get_parameters(free_only=False)
+    [0.56, 0.543, 0.133]
+
+    Trainable parameters are taken into account when calculating the Jacobian,
+    avoiding unnecessary calculations:
+
+    >>> tape.jacobian(dev)
+    [[-0.45478169]]
+    """
+
+    cast = staticmethod(np.array)
+
+    def __init__(self, name=None):
+        super().__init__()
+        self.name = name
+
+        self._prep = []
+        """list[~.Operation]: Tape state preparations."""
+
+        self._ops = []
+        """list[~.Operation]: quantum operations recorded by the tape."""
+
+        self._obs = []
+        """list[tuple[~.MeasurementProcess, ~.Observable]]: measurement processes and
+        coresponding observables recorded by the tape."""
+
+        self._par_info = {}
+        """dict[int, dict[str, Operation or int]]: Parameter information. Keys are
+        parameter indices (in the order they appear on the tape), and values are a
+        dictionary containing the corresponding operation and operation parameter index."""
+
+        self._trainable_params = set()
+        self._graph = None
+        self._output_dim = 0
+
+        self.wires = qml.wires.Wires([])
+        self.num_wires = 0
+
+        self.jacobian_options = {}
+
+        self.hash = 0
+        self.is_sampled = False
+
+        self._stack = None
```

Comment:
these are for backward compatibility, since devices currently expect these attributes to exist!